packages/insert-menu/src/InsertMenu.tsx

```diff
@@ -0,0 +1,302 @@
+import {SearchIcon, ThLargeIcon, UlistIcon} from '@sanity/icons'
+import {type SchemaType} from '@sanity/types'
+import {
+  Box,
+  Button,
+  Flex,
+  Grid,
+  Menu,
+  MenuItem,
+  type MenuItemProps,
+  Stack,
+  Tab,
+  TabList,
+  Text,
+  TextInput,
+  Tooltip,
+} from '@sanity/ui'
+import {type ChangeEvent, createElement, useReducer, useState} from 'react'
+import {isValidElementType} from 'react-is'
+
+import {getSchemaTypeIcon} from './getSchemaTypeIcon'
+import type {InsertMenuOptions} from './InsertMenuOptions'
+
+type InsertMenuGroup = NonNullable<InsertMenuOptions['groups']>[number] & {selected: boolean}
+type InsertMenuViews = NonNullable<InsertMenuOptions['views']>
+type InsertMenuView = InsertMenuViews[number]
+
+type InsertMenuEvent =
+  | {type: 'toggle view'; name: InsertMenuView}
+  | {type: 'change query'; query: string}
+  | {type: 'select group'; name: string | undefined}
+
+type InsertMenuState = {
+  query: string
+  groups: Array<InsertMenuGroup>
+  views: Array<{name: InsertMenuViews[number]; selected: boolean}>
+}
+
+function fullInsertMenuReducer(state: InsertMenuState, event: InsertMenuEvent): InsertMenuState {
+  return {
+    query: event.type === 'change query' ? event.query : state.query,
+    groups:
+      event.type === 'select group'
+        ? state.groups.map((group) => ({...group, selected: event.name === group.name}))
+        : state.groups,
+    views:
+      event.type === 'toggle view'
+        ? state.views.map((view) => ({...view, selected: event.name === view.name}))
+        : state.views,
+  }
+}
+
+const ALL_ITEMS_GROUP_NAME = 'all-items'
+
+const viewContainer = {
+  grid: {
+    component: Grid,
+    props: {autoRows: 'max', columns: 3, gap: 1},
+  },
+  list: {
+    component: Stack,
+    props: {space: 1},
+  },
+} as const
+
+/** @alpha */
+export type InsertMenuProps = InsertMenuOptions & {
+  schemaTypes: Array<SchemaType>
+  onSelect: (schemaType: SchemaType) => void
+  labels: {
+    'insert-menu.filter.all-items': string
+    'insert-menu.search.no-results': string
+    'insert-menu.search.placeholder': string
+    'insert-menu.toggle-grid-view.tooltip': string
+    'insert-menu.toggle-list-view.tooltip': string
+  }
```

Comment:
@stipsan Is this somewhat OK for labels? On the call site these are populated using translatable strings: https://github.com/sanity-io/sanity/pull/6853/files#diff-0f3eb80086c6f4701bf8c9bf9f95cee729761baf5a8a1dcb75e375b63d3ed25dR141-R147

Reply:
Yeah, nice and pragmatic 💖